src/boss/worker.go

```diff
@@ -2,214 +2,484 @@ package boss
 
 import (
 	"fmt"
+	"io"
 	"log"
-	"os"
-	"path/filepath"
-	"strings"
 	"net/http"
+	"os"
+	"os/exec"
+	"os/user"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type WorkerState int
+
+const (
+	STARTING WorkerState = iota
+	RUNNING
+	CLEANING
+	DESTROYING
 )
 
-type Invocation struct {
-	w    http.ResponseWriter
-	r    *http.Request
-	Done chan bool
+var (
+	clusterLogFile *os.File
+	taskLogFile    *os.File
+	clusterLog     *log.Logger
+	taskLog        *log.Logger
+	totalTask      int32
+	sumLatency     int64
+	nLatency       int64
+)
+
+type WorkerPool struct {
+	nextId  int
+	target  int
+	workers []map[string]*Worker
+	queue   chan *Worker
+	WorkerPoolPlatform
+	Scaling
+	sync.Mutex
```

Comment:
The lock is acquired before making changes to either target or workers (e.g. updateCluster()) to prevent multiple threads making changes to the cluster at the same time.